src/ucp/rndv/rndv_ppln.c

```diff
@@ -116,6 +125,95 @@ ucp_proto_rndv_ppln_init(const ucp_proto_init_params_t *init_params)
     return UCS_OK;
 }
 
+static void
+ucp_proto_rndv_ppln_frag_complete(ucp_request_t *freq, int send_ack,
+                                  ucp_proto_complete_cb_t complete_func,
+                                  const char *title)
+{
+    ucp_request_t *req = ucp_request_get_super(freq);
+
+    req->send.rndv.ppln.send_ack |= send_ack;
+    if (!ucp_proto_rndv_frag_complete(req, freq, title)) {
+        return;
+    }
+
+    if (req->send.rndv.rkey != NULL) {
+        ucp_proto_rndv_rkey_destroy(req);
+    }
+
+    if (req->send.rndv.ppln.send_ack) {
+        ucp_proto_request_set_stage(req, UCP_PROTO_RNDV_PPLN_STAGE_ACK);
+        ucp_request_send(req);
+    } else {
+        complete_func(req);
+    }
+}
+
+void ucp_proto_rndv_ppln_send_frag_complete(ucp_request_t *freq, int send_ack)
+{
+    ucp_proto_rndv_ppln_frag_complete(freq, send_ack,
+                                      ucp_proto_request_complete_success,
+                                      "ppln_send");
+}
+
+void ucp_proto_rndv_ppln_recv_frag_complete(ucp_request_t *freq, int send_ack)
+{
+    ucp_proto_rndv_ppln_frag_complete(freq, send_ack,
+                                      ucp_proto_rndv_recv_complete,
+                                      "ppln_recv");
+}
+
+static ucs_status_t ucp_proto_rndv_ppln_progress(uct_pending_req_t *uct_req)
+{
+    ucp_request_t *req  = ucs_container_of(uct_req, ucp_request_t, send.uct);
+    ucp_worker_h worker = req->send.ep->worker;
+    const ucp_proto_rndv_ppln_priv_t *rpriv;
+    ucp_datatype_iter_t next_iter;
+    ucs_status_t status;
+    ucp_request_t *freq;
+    uint8_t sg_count;
+
+    /* Nested pipeline is prevented during protocol selection */
+    ucs_assert(!(req->flags & UCP_REQUEST_FLAG_RNDV_FRAG));
+
+    req->send.state.completed_size = 0;
+    req->send.rndv.ppln.send_ack   = 0;
+    rpriv                          = req->send.proto_config->priv;
```

Comment:
minor: can be init in-place

Reply:
then clang-format will not align by "=" properly because this line would be too long